Analytic data for scalar wave in Kerr spacetime

prayush opened the pull request:
## Proposed changes

Analytic data for scalar wave in Kerr spacetime that can be used as initial data for starting a CurvedScalarWave evolution. This is templated on both the spacetime background and the flat-space scalar field profile (which can c.f. any class in `ScalarWave::Solutions`).

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+// Distributed under the MIT License.
```
Should these maybe go into a `CurvedWaveEquation` directory instead of just `WaveEquation`?

prayush replied:
Yes thats probably better going forward. I've moved all relevant code..

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+ * a Kerr black hole providing the spacetime background
+ *
+ * \details Here we obtain the flat-spacetime scalar-wave system's solution
+ * from `ScalarWave::Solutions::RegularSphericalWave` and suitably augment the
```
I would find it very helpful if the initial data were actually written out for the scalar field as an equation.

prayush replied:
Added

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+  static constexpr OptionString help{
+      "A regular spherical wave solution of the covariant wave equation"
+      " with a Kerr black hole (in Kerr-Schild coordinates) as background.\n"
+      "WARNING: This solution should only be used to set initial data."};
```
I'm confused now. Is this a solution or data?

prayush replied:
This is absolutely "analytic data" and anything that suggests otherwise needs to be changed. I've clarified the phrasing here as well as have written out the source equations explicitly above.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+      "A regular spherical wave solution of the covariant wave equation"
+      " with a Kerr black hole (in Kerr-Schild coordinates) as background.\n"
+      "WARNING: This solution should only be used to set initial data."};
+  static std::string name() noexcept { return "RegularSphWaveKerr"; };
```
Is `SphericalWaveKerr` short enough? I don't think the `Regular` is necessary since if we evolved a singular wave things would be very bad :man_shrugging:

prayush replied:
Sure, I've renamed it.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+  using tags = tmpl::append<spacetime_tags, scalar_wave_tags>;
+
+  // Returns only U
+  tuples::tagged_tuple_from_typelist<scalar_wave_tags> variables(
```
This is different from the implementations we normally use where you can get any tag. Please change the interface to allow only retrieving a single tag. Once #1893 this should be even easier.

prayush replied:
Indeed, but I deliberately did this here since this is analytic initial data and not a solution, and it should only be used to set the initial values of `{psi, Phi, Pi}` together.

Can there be a scenario where only a subset of initial variables are needed to be set? E.g. only `psi` or only `Pi` or only the `lapse`? Even if they are needed wont getting them from AnalyticSolutions be a better idea anyway..?

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+  gr::Solutions::KerrSchild kerr_schild_soln_;
+};
+
+SPECTRE_ALWAYS_INLINE bool operator==(
```
Please make `operator==` a friend and then define it in the `cpp` file.

prayush replied:
done

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,173 @@
+  }
+
+ private:
+  double mass_{1.};
```
Is there a reason not to make the default values signaling NaN? I'm worried that having the defaults here can cause issues with serialization that are difficult to track down. Since I don't expect we will be default constructing analytic data very often I think the default contsructor is only there for Charm++, really.

prayush replied:
I agree that it would be cleaner to default these options to NaNs. However, the routine `validate_default()` invoked in `ParseOptions.hpp` throws an FPE as it tries to assert that `default_value() > lower_bound()` for input options, when either of the two is NaN.. Perhaps there is a way to disable this check, but I am not aware of it. Is there?

prayush replied:
changed to NaNs. (Sorry I seem to have confused myself between `default_value()` and the default-construction values..)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,129 @@
+namespace AnalyticData {
+
+RegularSphericalWaveKerrSchild::RegularSphericalWaveKerrSchild(
+    double mass,
```
I think the first 3 args can be `const` here

prayush replied:
Made `const`

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,129 @@
+    double mass,
+    RegularSphericalWaveKerrSchild::BlackHoleSpin::type dimensionless_spin,
+    RegularSphericalWaveKerrSchild::BlackHoleCenter::type center,
+    std::unique_ptr<MathFunction<1>> profile, const OptionContext& context)
```
Since you don't use the `OptionContext` I don't think you need it as an argument in the constructor at all, if I remember correctly.

prayush replied:
I **do** use it when constructing the KerrSchild solution below ..

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,129 @@
+tuples::tagged_tuple_from_typelist<
+    RegularSphericalWaveKerrSchild::scalar_wave_tags>
+RegularSphericalWaveKerrSchild::variables(
+    const tnsr::I<DataVector, volume_dim>& x, const double t,
```
`AnalyticData` should not take `t` as an argument at all...

prayush replied:
Of course, this is an oversight. I've remedied it now but removing time input everywhere

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,129 @@
+  get<Psi>(result) = get<ScalarWave::Psi>(flat_space_scalar_wave_vars);
+  get<Phi<volume_dim>>(result) =
+      get<ScalarWave::Phi<volume_dim>>(flat_space_scalar_wave_vars);
+  // finally, get PI
```
PI or pi or Pi?

prayush replied:
disambiguated

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,129 @@
+      get<ScalarWave::Phi<volume_dim>>(flat_space_scalar_wave_vars);
+  // finally, get PI
+  {
+    // Which dt<Psi> do we use in the definition of Pi?
```
Did you find the answer to this?

prayush replied:
yes, its the coordinate time derivative. Added this iinfo to the documentation fo the struct, where it should be.

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/WaveEquation/Test_RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,247 @@
+      "A regular spherical wave in Kerr spacetime"};
+};
+
+void test_construct_from_options() noexcept {
```
Could you run the result of the construction and serialization through a full test? Also, let's avoid 1 and 0 for values as read in.

prayush replied:
done

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/WaveEquation/Test_RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,247 @@
+      1.0, {{0.1, 0.2, 0.3}}, {{0.0, 0.0, 0.0}},
+      std::make_unique<MathFunctions::Gaussian>(1., 1., 0.));
+  Parallel::register_derived_classes_with_charm<MathFunction<1>>();
+  test_serialization(solution);
```
Here too, pass the deserialized solution through a full test of all the values.

prayush replied:
Added such a test, please have a look and see if this is what you were thinking of..

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/WaveEquation/Test_RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,247 @@
+  const std::array<double, 3> center{{0.3, 0.2, 0.4}};
+  const CurvedScalarWave::AnalyticData::RegularSphericalWaveKerrSchild solution{
+      mass, spin, center,
+      std::make_unique<MathFunctions::Gaussian>(1., 1., 0.)};
```
Can we use non-unit and non-zero values?

prayush replied:
changed everywhere

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/WaveEquation/Test_RegularSphericalWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,247 @@
+                        get<ScalarWave::Phi<3>>(flat_vars).get(2));
+}
+
+void test_kerr() noexcept {
```
Can you add a python test for this?

prayush replied:
For a python test of this `AnalyticData` I would need to write up both the kerr-schild solution & flat-space regular spherical wave in python, since those solutions don't already have python implementations that they are being tested against. Is it not better if I add it as a separate PR(s) and not delay this one, because I can use their python implementations to add tests for kerr-schild and flat-space spherical wave solutions themselves in the same go then?

nilsdeppe replied:
Sounds good to me

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/WaveEquation/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+  ${LIBRARY}
+  "PointwiseFunctions/AnalyticData/WaveEquation/"
+  "${LIBRARY_SOURCES}"
+  "GeneralRelativitySolutions;GeneralRelativity;WaveEquation;WaveEquationAnalyticData;MathFunctions;Utilities"
```
alphabet

prayush replied:
fixed

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/WaveEquation/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+  ${LIBRARY}
+  INTERFACE DataStructures
+  INTERFACE ErrorHandling
+  INTERFACE GeneralRelativity
```
Spectral, Parallel, Utilities

prayush replied:
Included Spectral and Utilities, but which library do you mean by `Parallel`? Also, why does this need to link with Spectral?

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -56,20 +57,30 @@ class RegularSphericalWaveKerrSchild : public MarkAsAnalyticData {
     static constexpr OptionString help = {"Mass of the black hole"};
-    static type default_value() noexcept { return 1.; }
+    static type default_value() noexcept {
+      return std::numeric_limits<double>::signaling_NaN();
```
no, don't change the defaults in the option structs...  just the default initialization of the member variables...

prayush replied:
That sounds like an inconsistent thing to do ... unless the `default_value()` is used when the user does not specify any value in the input file (which is different from default constructing the solution)..

kidder replied:
yes, that is how `default_value()` is used...

prayush replied:
great, thanks for the explanation. I've changed to NaN's.

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/RegularSphericalWaveKerrSchild.hpp:
```diff
@@ -74,10 +102,11 @@ class RegularSphericalWaveKerrSchild : public MarkAsAnalyticData {
-      " with a Kerr black hole (in Kerr-Schild coordinates) as background.\n"
-      "WARNING: This solution should only be used to set initial data."};
-  static std::string name() noexcept { return "RegularSphWaveKerr"; };
+      "A regular spherical wave solution of the covariant wave equation\n"
```
"A regular spherical wave with a Kerr..."  (i.e. omit solution of ... equation) .   And you can omit the WARNING as it is marked as analytic data

prayush replied:
changed

prayush commented:
@nilsdeppe @kidder : I've templated this analytic data class on flat-space scalar field solutions instead of hard-coding for a `RegularSphericalWave`. Please do have a look when you have a bit of time, thanks!

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,213 @@
+                                Dummy>::value>>
+ScalarWaveKerrSchild<ScalarWave::Solutions::RegularSphericalWave>::
+    ScalarWaveKerrSchild(const double mass,
+                         std::array<double, volume_dim> dimensionless_spin,
```
make all the arguments const except `profile` and don't bother with the `move` below. They're just `double`s so copy is cheap and move is impossible. I know we've been inconsistent about this in the past, and that's partly because compilers and clang-tidy were not as good about warning about incorrect things done in a constructor as they are now.

prayush replied:
Changed all to const, removed `move`s for all (except profile)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,213 @@
+    Requires<std::is_same<ScalarWave::Solutions::PlaneWave<3>, Dummy>::value>>
+ScalarWaveKerrSchild<ScalarWave::Solutions::PlaneWave<3>>::ScalarWaveKerrSchild(
+    const double mass, std::array<double, volume_dim> dimensionless_spin,
+    std::array<double, volume_dim> center,
```
same comment about copy/move/const

prayush replied:
Done as above

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+  using tags = tmpl::append<spacetime_tags, scalar_wave_tags>;
+
+  /// Returns curved scalar wave variables
+  tuples::tagged_tuple_from_typelist<scalar_wave_tags> variables(
```
can you update the interface to conform with the way the interfaces work for all our other analytic solutions? Will recently updated Kerr-Schild to conform to the standard way and I'd rather not have new ones that need to be updated in the future since that leads to people copying the incorrect interface.

prayush replied:
Updated the interface, now any arbitrary number / combination of tags involved can be retrieved. This is cleaner too.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+ * \psi &:= \overset{\circ}{\psi} \\
+ * \Phi_i &:= \partial_i\psi = \overset{\circ}{\partial_i\psi}
+ *                           = \overset{\circ}{\Phi_i} \\
+ * \Pi &:= -t^a \partial_a \psi = \frac{1}{\alpha}
```
`t^a` -> `n^a`

prayush replied:
renamed

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+      const ScalarWaveKerrSchild<LocalScalarFieldSolution>& rhs) noexcept;
+
+  template <typename LocalScalarFieldSolution>
+  friend bool operator!=(
```
`!=` does not need to be a `friend` because it just calls `not(==)`

prayush replied:
`!=` changed to nonmember overload

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,377 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
```
these will need an update

prayush replied:
updated

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveKerrSchild.cpp:
```diff
@@ -0,0 +1,377 @@
+        {{0.0, 0.0, 0.0}},
+        std::make_unique<MathFunctions::Gaussian>(11., 1.4, 0.3)};
+
+    test_kerr(curved_wave_data_constructed_from_opts, bh_solution,
```
can you do what the Kerr-Schild solution does with `TestHelpers::AnalyticSolutions::test_tag_retrieval` to make sure the interface works correctly?

prayush replied:
Added tests for tag retrieval. Had to add an equivalent `TestHelpers::AnalyticData::test_tag_retrieval` for analytic data.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+// IWYU pragma: no_include <pup.h>
+
+namespace CurvedScalarWave {
+namespace AnalyticData {
```
Can you add this namespace to the `AnalyticData` group so it shows up in the documentation?

prayush replied:
Added

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+ *
+ * where
+ *
+ * - the hollow dots \f$\overset{\circ}{}\f$ over quantities indicate that they
```
I find this circle notation a bit cumbersome. The only quantity that changes is `\Pi`, so I would just not have the circles and define `\Pi` the way you do above. The definitions are straight from the curved scalar wave system itself, so I'm not sure that trying to differentiate things with the circles clarifies anything.

prayush replied:
Simplified the description

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,225 @@
+ * - \f$t^a\f$ is the unit normal to spatial slices that spacetime has been
+ * foliated into.
+ *
+ * That the coordinate profile of scalar field \f$\psi\f$ in curved spacetime is
```
I don't think this level of detail is needed in an `analytic data` since this is just the definitions from the curved scalar wave system itself. If you want this explained somewhere, I feel like it should be on the CSW system itself, rather than an analytic data.

prayush replied:
Simplified the description

nilsdeppe reviewed (changes requested):
Since they'll be fairly major changes, just squash right away.

kidder commented on src/PointwiseFunctions/AnalyticData/CMakeLists.txt:
```diff
@@ -3,3 +3,4 @@
 
 add_subdirectory(Burgers)
 add_subdirectory(GrMhd)
+add_subdirectory(CurvedWaveEquation)
```
alphabet

prayush replied:
fixed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,230 @@
+ * this class provides methods to initialize the scalar-field and spacetime
+ * variables using analytic solutions to the flat-space scalar-wave equation
+ * and a Kerr background spacetime. That the coordinate profile of the scalar
+ * field \f$\psi\f$ in curved spacetime is the same as that in flat spacetime
```
use capital Psi

prayush replied:
changed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,230 @@
+ * future-directed time derivative of the scalar field:
+ *
+ * \f{align}
+ * \Pi := -n^a \partial_a \psi
```
use alpha and beta for lapse and shift

prayush replied:
changed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveKerrSchild.hpp:
```diff
@@ -0,0 +1,230 @@
+ * where \f$n^a\f$ is the unit normal to spatial slices of the spacetime
+ * foliation.
+ */
+template <typename ScalarFieldSolution>
```
Let's template on both the ScalarFieldData and the BackgroundGrData so...
```
template <typename ScalarFieldData, typename BackgroundGrData>
class OnBackground : public MarkAsAnalyticData {
public:
struct ScalarField {
 using type = ScalarFieldData;
};
struct Background {
 using type = BackgroundGrData;
};
...
private:
 ScalarFieldData scalar_field_data_;
 BackgroundGrData background_gr_data_;
};\
```

prayush replied:
done

kidder reviewed (changes requested):
reviewed with @nilsdeppe

nilsdeppe commented:
I'm happy for you to squash the fixup

prayush committed:
Add test helper for analytic data

which checks that tags provided by the data
can be retrieved both individually, and all
at once

prayush commented:
Hi @sxs-collaboration/spectre-core-devs - all review comments have been addressed, including templating this analytic data class on both the background spacetime and the scalar field solution. Its ready for another look / merge.

kidder commented on src/PointwiseFunctions/AnalyticData/CMakeLists.txt:
```diff
@@ -2,5 +2,6 @@
 # See LICENSE.txt for details.
```
remove the caveat from the commit message as we do not support non-self starting time steppers any more.

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+ * \details When evolving a scalar field propagating through curved spacetime,
+ * this class provides methods to initialize the scalar-field and spacetime
+ * variables using analytic solution(s) of the flat-space scalar-wave equation
+ * and of Einstein equations. That the coordinate profile of the scalar
```
the Einstein 

 Note that the cooordinate...

prayush replied:
changed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+ * field \f$\Psi\f$ in curved spacetime is the same as that in flat spacetime
+ * is our primary identification, allowing it to be initialized using any
+ * member class of `ScalarWave::Solutions`. Solving constraints of the
+ * `CurvedScalarWave` system, we also identify the coordinate spatial derivative
```
Simply state that we initialize Phi to the partial derivative of Psi

prayush replied:
changed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+template <typename ScalarFieldData, typename BackgroundGrData>
+class ScalarWaveGr : public MarkAsAnalyticData {
+ public:
+  static constexpr size_t volume_dim = 3;
```
why restrict to 3?  Just make sure the two template arguments have the same `volume_dim` and use that

prayush replied:
Updated

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+  static std::string name() noexcept { return "ScalarWaveGr"; };
+
+  // Construct from options
+  explicit ScalarWaveGr(BackgroundGrData&& background,
```
drop explicit and take the arguments by value

prayush replied:
updated

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+      : flat_space_scalar_wave_data_(std::move(scalar_field)),
+        background_gr_data_(std::move(background)) {}
+
+  // Enable or disable this constructor based on ScalarFieldData
```
remove this constructor

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+      : flat_space_scalar_wave_data_(std::move(profile)),
+        background_gr_data_(mass, dimensionless_spin, center, context) {}
+
+  // Enable or disable this constructor based on ScalarFieldData
```
remove this constructor also

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+    return {std::move(
+        get<ScalarWave::Phi<volume_dim>>(flat_space_scalar_wave_data_.variables(
+            x, default_initial_time,
+            tmpl::list<ScalarWave::Pi, ScalarWave::Phi<volume_dim>,
```
just use a `tmpl::list<ScalarWave::Phi<volume_dim>>`

prayush replied:
[`PlaneWave`](https://github.com/sxs-collaboration/spectre/blob/develop/src/PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp) and [`RegularSphericalWave`](https://github.com/sxs-collaboration/spectre/blob/develop/src/PointwiseFunctions/AnalyticSolutions/WaveEquation/RegularSphericalWave.hpp) dont yet have the overloads of `variables` needed to fetch one tag at a time. I am happy to fix that in a follow-up PR to this.

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+  /// Retrieve scalar wave variables
+  tuples::TaggedTuple<Pi> variables(const tnsr::I<DataVector, volume_dim>& x,
+                                    tmpl::list<Pi> /*meta*/) const noexcept;
+  tuples::TaggedTuple<Phi<volume_dim>> variables(
```
add a blank line

prayush replied:
added

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+    return {
+        std::move(get<ScalarWave::Psi>(flat_space_scalar_wave_data_.variables(
+            x, default_initial_time,
+            tmpl::list<ScalarWave::Pi, ScalarWave::Phi<volume_dim>,
```
just use `tmpl::list<ScalarWave::Psi>`

prayush replied:
please see [above](https://github.com/sxs-collaboration/spectre/pull/1832#discussion_r427766722)

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/AnalyticData/AnalyticData.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
```
don't need the includes for specific solutions

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,203 @@
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/RegularSphericalWave.hpp"
+#include "PointwiseFunctions/GeneralRelativity/TagsDeclarations.hpp"
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
```
don't need this

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,145 @@
+      x, default_initial_time, spacetime_tags<DataVector>{});
+  auto result = make_with_value<tuples::TaggedTuple<Pi>>(x, 0.);
+
+  {
```
don't need this pair of braces

prayush replied:
removed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,145 @@
+          x, default_initial_time,
+          tmpl::list<ScalarWave::Pi, ScalarWave::Phi<volume_dim>,
+                     ScalarWave::Psi>{});
+  const auto flat_space_dt_scalar_wave_vars =
```
don't use the time derivatives; just use Pi from the flat-space solution/data

prayush replied:
changed

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,145 @@
+
+#include <pup.h>
+
+#include "DataStructures/DataBox/Prefixes.hpp"
```
double-check these includes, you don't need a lot of them

prayush replied:
removed relic includes, thanks!

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,145 @@
+#undef STYPE
+
+}  // namespace CurvedScalarWave::AnalyticData
+/// \endcond
```
turn this file into a `.tpp` file dropping the instantiations.  Then for each combination of ScalarWave/Background create a simple `.cpp` and `.hpp` pair where the `.cpp` file has the instantiation and the `.hpp` file just has includes of the specific flat-space solution, background solution and the `tpp` file

prayush replied:
updated

kidder commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,402 @@
+  const auto kerr_variables = bh_solution.variables(
+      x, 0., gr::Solutions::KerrSchild::tags<DataVector>{});
+
+  // construct the vars in-situ
```
construct the expected vars

prayush replied:
updated

kidder commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,402 @@
+        get<gr::Tags::Shift<3, Frame::Inertial, DataVector>>(kerr_variables),
+        get<ScalarWave::Phi<3>>(flat_wave_vars));
+    get(local_pi) = (get(shift_dot_dpsi) -
+                     get(get<::Tags::dt<ScalarWave::Psi>>(flat_wave_dt_vars))) /
```
don't use the time derivative, it will get removed eventually

prayush replied:
updated, now using `Pi`.

kidder reviewed (changes requested):
reviewed with @nilsdeppe

prayush commented:
> reviewed with @nilsdeppe

Hi @kidder - this has again been updated, and is ready for another look / merge!

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/RegularSphericalWave3DKerrSchild.cpp:
```diff
@@ -0,0 +1,30 @@
+#define STYPE(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define BTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
+
+#define INSTANTIATE(_, data)                                       \
```
just instantiate directly instead of using macros

kidder commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWave3DKerrSchild.cpp:
```diff
@@ -0,0 +1,29 @@
+#define STYPE(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define BTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
+
+#define INSTANTIATE(_, data)                                       \
```
just instantiate directly instead of using macros

kidder reviewed (changes requested):
you can squash including these changes

prayush committed:
Add volume_dim to analytic solutions of the ScalarWave system

prayush committed:
AnalyticData for scalar waves in curved spacetime

This data is constructed using solutions of the scalar wave
equation in flat spacetime, augmented to account for a
curved background geometry. It can be used to combine
any of available ScalarWave and GeneralRelativity solutions
to provide initial data for evolving CurvedScalarWave
system.

prayush commented:
Hi @sxs-collaboration/spectre-core-devs : this is probably ready to merge.

nilsdeppe commented:
Why did this get squashed? Now I have to review the whole thing again... That'll take me until next week, sorry.

kidder commented:
My apologies @nilsdeppe @prayush this was my fault, I misread the comment by Prayush above as implying that you two had looked at the previous version together and told Prayush to squash.

prayush commented:
No worries, this can wait a few days @nilsdeppe !

prayush commented:
Hi @nilsdeppe / @sxs-collaboration/spectre-core-devs - this PR is ready for a final pass / merge. Whenever you find a bit of time, please have a look!

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:
```diff
@@ -0,0 +1,164 @@
+// See LICENSE.txt for details.
+
+#pragma once
+
```
need `cstddef`, `string`, `utility`

prayush replied:
added

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:
```diff
@@ -0,0 +1,409 @@
+#include "Framework/TestingFramework.hpp"
+
+#include <array>
+#include <memory>
```
and `utility` for `std::move`

prayush replied:
added

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/RegularSphericalWave3DKerrSchild.cpp:
```diff
@@ -0,0 +1,21 @@
+
+/// \cond
+namespace CurvedScalarWave ::AnalyticData {
+template class ScalarWaveGr<ScalarWave::Solutions::RegularSphericalWave,
```
Not a change request: Ideally we'd like to document what instantiations we have, but I don't really know of a good way to do that other than manually keeping track of it in the class template documentation. Just a note in case anyone comes up with ideas in the future :)

nilsdeppe reviewed (commented):
Looks good, a few missing includes. Would you mind adding them and rebasing on develop? I'm happy with this :)

nilsdeppe reviewed (approved):
LGTM! Thanks for implementing this and sorry it took so long to review :(

prayush commented:
ping @sxs-collaboration/spectre-core-devs - this is good to go!
